src/arithmetic/comprehension_funcs/comprehension_funcs.c

```diff
@@ -157,12 +170,17 @@ SIValue AR_SINGLE(SIValue *argv, int argc, void *private_data) {
 		Record_AddScalar(r, ctx->variable_idx, current_elem);
 
 		// If more then 1 element in a SINGLE function pass the predicate, return false.
-		if(FilterTree_applyFilters(ctx->ft, r)) {
+		FT_Result res = FilterTree_applyFilters(ctx->ft, r);
+		if(res == FILTER_PASS) {
 			if(single) return SI_BoolVal(false);
 			else single = true;
 		}
+		is_null |= (res == FILTER_NULL);
```

Comment:
Perfect!
In the case of SINGLE we can't return early
```cypher
RETURN single(x IN [1,null,3,1] WHERE x=1) AS result
```

Should return `False`